src/ossl/ecc.rs

```diff
@@ -709,3 +716,149 @@ impl Verify for EccOperation {
+                if max == CK_UNAVAILABLE_INFORMATION as usize {
+                    return err_rv!(CKR_MECHANISM_INVALID);
+                }
+                match factory.as_secret_key_factory()?.recommend_key_size(max) {
+                    Ok(len) => len,
+                    Err(_) => return err_rv!(CKR_TEMPLATE_INCONSISTENT),
+                }
+            }
+        };
+        /* these do not apply to the raw ECDH */
+        if self.kdf != CKD_NULL {
+            params = params
+                .add_const_c_string(
+                    name_as_char(OSSL_EXCHANGE_PARAM_KDF_TYPE),
+                    ckd_to_type_name(self.kdf),
+                )?
+                .add_const_c_string(
+                    name_as_char(OSSL_EXCHANGE_PARAM_KDF_DIGEST),
+                    mech_type_to_digest_name(
+                        self.hash.ok_or(to_rv!(CKR_MECHANISM_INVALID))?,
+                    ),
+                )?;
+            if self.shared.len() > 0 {
+                params = params.add_octet_string(
+                    name_as_char(OSSL_EXCHANGE_PARAM_KDF_UKM),
+                    &self.shared.to_vec(),
+                )?;
+            }
+            params = params.add_uint(
+                name_as_char(OSSL_EXCHANGE_PARAM_KDF_OUTLEN),
+                keylen as u32,
+            )?;
+        }
+
+        params = params.finalize();
+
+        let mut pkey = object_to_ecc_private_key(key)?;
+        let mut ctx = new_pkey_ctx_from_obj(&mut pkey)?;
+        let res = unsafe {
+            EVP_PKEY_derive_init_ex(ctx.as_mut_ptr(), params.as_ptr())
+        };
+        if res != 1 {
+            return err_rv!(CKR_DEVICE_ERROR);
+        }
+
+        /* Import peer key */
+        let curve_name = get_curve_name_from_obj(key)?;
+        let public = self.public.to_vec();
+        let mut peer = make_ecc_public_key(curve_name, public)?;
+
+        let res = unsafe {
+            EVP_PKEY_derive_set_peer(ctx.as_mut_ptr(), peer.as_mut_ptr())
+        };
+        if res != 1 {
+            return err_rv!(CKR_DEVICE_ERROR);
+        }
+
+        let mut secret_len = 0usize;
+        let res = unsafe {
+            EVP_PKEY_derive(
+                ctx.as_mut_ptr(),
+                std::ptr::null_mut(),
+                &mut secret_len,
+            )
+        };
+        if res != 1 {
+            return err_rv!(CKR_DEVICE_ERROR);
+        }
+        let mut secret = vec![0u8; secret_len];
+        if secret_len < keylen {
+            return err_rv!(CKR_TEMPLATE_INCONSISTENT);
+        }
+        let res = unsafe {
+            EVP_PKEY_derive(
+                ctx.as_mut_ptr(),
+                secret.as_mut_ptr(),
+                &mut secret_len,
+            )
+        };
+        if res != 1 {
+            return err_rv!(CKR_DEVICE_ERROR);
+        }
+
+        let mut tmpl = template.to_vec();
+
+        tmpl.push(CK_ATTRIBUTE::from_slice(
+            CKA_VALUE,
+            &secret[(secret_len - keylen)..],
```

Comment:
It is surprising that you have to pick the tail of the buffer ...

Reply:
The pkcs11 way if I understand it right: 

> This mechanism derives a secret value, and truncates the result according to the CKA_KEY_TYPE attribute of the template and, if it has one and the key type supports it, the CKA_VALUE_LEN attribute of the template.  (**The truncation removes bytes from the leading end of the secret value.**)  The mechanism contributes the result as the CKA_VALUE attribute of the new key; other attributes required by the key type must be specified in the template.

https://docs.oasis-open.org/pkcs11/pkcs11-curr/v3.0/os/pkcs11-curr-v3.0-os.html#_Toc30061194

Reply:
is this true for all cases or only for the the case where no hashing is used?

Reply:
I read it as in all cases.

Reply:
This applies to the raw ECDH only, which has fixed-length output based on the field length. The X9.63 (with hashing kdf) requires the length as an input already so no truncation is happening there (as `secret_len = keylen` here). I might make it more explicit if you want.